main.cpp

```diff
@@ -170,11 +328,11 @@ int main()
     Numeric<float> floatNum(4.3f);
     Numeric<int> intNum(2);
     Numeric<int> intNum2(6);
-    intNum = 2 + (intNum2 - 4) + floatNum / 2.3;
+    intNum = 2 + (intNum2 - 4) + static_cast<double>(floatNum) / 2.3;
     std::cout << "intNum: " << intNum << std::endl;
     
     {
-        using Type = decltype(f)::Type;
+        using Type = decltype(f)::CurrentType;
         f.apply([&f](std::unique_ptr<Type>&value) -> decltype(f)&
```

Comment:
This works.  but you could have done: 
```
using Type = decltype(f);
f.apply([&f](std::unique_ptr<Type::Type>& value) -> Type&
```
one less `decltype(f)`
- [ ] revise if you want to.